model/src/main/kotlin/ResolvedConfiguration.kt

```diff
@@ -28,10 +33,70 @@ import com.fasterxml.jackson.annotation.JsonInclude
  * TODO: Add further data.
  */
 data class ResolvedConfiguration(
+    @JsonInclude(value = JsonInclude.Include.CUSTOM, valueFilter = PackageCurationsFilter::class)
+    val packageCurations: PackageCurations = PackageCurations(),
+) {
+    /**
+     * Return all [PackageCuration]s contained in this [ResolvedConfiguration] in highest-priority-first order.
+     */
+    @JsonIgnore
+    fun getAllPackageCurations(): List<PackageCuration> =
+        packageCurations.providers.flatMap { id ->
+            packageCurations.data[id].orEmpty()
+        }
+}
+
+@JsonSerialize(converter = PackageCurationsConverter::class)
+data class PackageCurations(
     /**
-     * All package curations applicable to the packages contained in the enclosing [OrtResult] ordered
-     * highest-priority-first.
+     * All enabled providers ordered highest-priority-first.
      */
-    @JsonInclude(JsonInclude.Include.NON_EMPTY)
-    val packageCurations: List<PackageCuration> = emptyList()
-)
+    val providers: List<String> = emptyList(),
+
+    /**
+     * All package curations applicable to the packages contained in the enclosing [OrtResult].
+     */
+    val data: Map<String, Set<PackageCuration>> = emptyMap()
+) {
+    @JsonIgnore
+    fun isEmpty(): Boolean =
+        providers.isEmpty() && data.isEmpty()
+
+    init {
+        val duplicateProviders = providers.getDuplicates()
+        require(duplicateProviders.isEmpty()) {
+            "The list 'providers' contains the following duplicates, which is not allowed: " +
+                    "${duplicateProviders.joinToStringSingleQuoted()}."
+        }
+
+        val invalidProviderReferences = data.keys.filter { it !in providers }
+        require(invalidProviderReferences.isEmpty()) {
+            "The following keys in 'data' reference a non-existing provider, which is not allowed: " +
+                    "${invalidProviderReferences.joinToStringSingleQuoted()}."
+        }
+    }
+}
+
+@Suppress("EqualsOrHashCode", "EqualsWithHashCodeExist") // The class is not supposed to be used with hashing.
+private class PackageCurationsFilter {
+    override fun equals(other: Any?): Boolean =
+        other is PackageCurations && other.isEmpty()
+}
+
+/**
+ * Sorts the keys of [PackageCurations.data] alphabetically, the values of [PackageCurations.data] by
```

Comment:
> Sorts the keys of [PackageCurations.data] alphabetically

Should we really do this? Wouldn't it be more natural to simply use the same order as in `providers`?

> the values of [PackageCurations.data] by their [id][PackageCuration.id]

Similarly here, even if we agreed that the order of curations within a single provider should not matter, shouldn't we simply maintain the order in which the provider returned them? This would easy debugging if we ever ran into the case where the order of curations from a single provider *does* matter.

> drops entries for providers without curations from [PackageCurations.data].

IMO those should already be filtered out in `resolveConfiguration()`.

Reply:
> Should we really do this? Wouldn't it be more natural to simply use the same order as in providers?

I believe there is value in having things ordered. In particular:

1. easier to find things
2. related things are next to on another

I'm not quite sure what you want to say with the term "more natural". Can you rephrase please?
Note that we agreed as part of our discussion on this PR that for a particular provider the order does not have meaning, so it is arbitrary. I don't see why arbitrary order is natural and better than sorted order.

Reply:
> for a particular provider the order does not have meaning

Right, but my sentence that you're quoting referred to "the keys of [PackageCurations.data]", i.e. the ids of the providers. I would expect the ids here to match the ids below `providers`. That what I refer to as "natural".

Reply:
> Right, but my sentence that you're quoting referred to "the keys of [PackageCurations.data]", i.e. the ids of the providers. I would expect the ids here to match the ids below `providers`. That what I refer to as "natural".

Oh, I see. I've got not that strong preference for the key ordering, but slightly prefer to keep the order as-is. Let's just also add this to our core devs discussion on it as the PR is blocked until then anyway .